### PR TITLE
HSEARCH-3052 Search 6 groundwork - Restore the Elasticsearch index schema management strategies (validation, ...)

### PR DESCRIPTION
**Pull request by yrodiere**
https://hibernate.atlassian.net//browse/HSEARCH-3052

~Based on #1865  and #1863 , which should be merged first.~ => Done

Note that I only did the minimum to get this feature in the next Alpha, as I think Search is pretty unusable without it. Some improvements should be considered, though: [HSEARCH-3468](https://hibernate.atlassian.net/browse/HSEARCH-3468), [HSEARCH-3467](https://hibernate.atlassian.net/projects/HSEARCH/issues/HSEARCH-3467) in particular.

### CODE REVIEW

File: backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/index/impl/ElasticsearchIndexManagerBuilder.java [hunk 1]
```diff
@@ -71,27 +76,31 @@ public ElasticsearchIndexManagerImpl build() {
...
 					indexingBackendContext, searchBackendContext,
 					hibernateSearchIndexName, encodedElasticsearchIndexName,
-					encodedTypeName,model,
+					encodedTypeName, model,
+					indexManagementStrategy,
 					serialOrchestrator, parallelOrchestrator,
 					refreshAfterWrite
 			);
+
+			// TODO perform index initialization in parallel for all indexes?
```
Comment (fax4ever) [inline backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/index/impl/ElasticsearchIndexManagerBuilder.java]: I think that `indexManager.start()` could be moved outside from the builder.
One option could be to move it to the constructor of `ElasticsearchIndexManagerImpl`, another option could be to move it to the builder client class.

File: backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/cfg/ElasticsearchIndexManagementStrategyConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+import java.util.Locale;
+import java.util.Map;
+
+/**
+ * Strategy for creating/deleting the indexes in Elasticsearch upon Hibernate Search initialization and shut-down.
+ *
+ * @author Gunnar Morling
+ * @author Yoann Rodiere
+ */
+public enum ElasticsearchIndexManagementStrategyConfiguration {
```
Comment (fax4ever) [inline backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/cfg/ElasticsearchIndexManagementStrategyConfiguration.java]: Maybe we can change this name is something more synthetic, such as `ElasticsearchIndexCreationStrategy`

File: backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/work/impl/CreateIndexWork.java [hunk 3]
```diff
@@ -68,9 +68,14 @@ public Builder settings(IndexSettings settings) {
 		public Builder mapping(URLEncodedString typeName, RootTypeMapping mapping) {
```
Comment (fax4ever) [inline backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/work/impl/CreateIndexWork.java]: I don't know actually if this method is used.

File: backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/index/admin/impl/ValidationContextType.java [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.search.backend.elasticsearch.index.admin.impl;
+
+enum ValidationContextType {
+
+	INDEX,
+	MAPPING,
```
Comment (fax4ever) [inline backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/index/admin/impl/ValidationContextType.java]: Maybe we could remove this enumeration value.

File: backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/index/management/impl/ElasticsearchIndexManagementStrategy.java [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.search.backend.elasticsearch.index.management.impl;
+
+import org.hibernate.search.backend.elasticsearch.cfg.ElasticsearchIndexManagementStrategyConfiguration;
+import org.hibernate.search.backend.elasticsearch.index.admin.impl.ElasticsearchIndexAdministrationClient;
+import org.hibernate.search.backend.elasticsearch.index.admin.impl.ElasticsearchIndexManagementExecutionOptions;
+import org.hibernate.search.util.AssertionFailure;
+
+public class ElasticsearchIndexManagementStrategy {
```
Comment (fax4ever) [inline backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/index/management/impl/ElasticsearchIndexManagementStrategy.java]: In my opinion, this is a very good use of the information holder pattern.

**Comment by yrodiere** [comment]
Jenkins, retest this please

**Comment by yrodiere** [comment]
Merged, thanks!

**Pull request merged**